### PR TITLE
xds_end2end_test: Move fault injection tests into their own file

### PR DESCRIPTION
**Pull request by markdroth**

### CODE REVIEW

**Review by lidizheng — APPROVED** [review]

File: test/cpp/end2end/xds/BUILD
Comment (markdroth) [reply test/cpp/end2end/xds/BUILD]: Good point, removed.

**Comment by markdroth** [comment]
The `XdsFaultInjectionMaxFault` test seems to be failing consistently in this PR.  I don't know what's changed; the test code is exactly the same as it was in the other file.  Can you please take a look and see why this PR is tickling it?

**Comment by lidizheng** [comment]
Okay, I will take a look later.

**Comment by lidizheng** [comment]
I can reproduce this problem locally. I observed that if I ran the entire fault injection test suite, all 4 MaxFault tests will fail. If I only ran the MaxFault tests, the first one will pass, other ones will fail.

After checking the global value of the global atomic `g_active_faults`, I found that the decrement logic were left out in the promise conversion. Specifically, the following method didn't get translated to the promise version (and two call sites) (https://github.com/grpc/grpc/pull/29288):

```c++
  void FaultInjectionFinished() {
    g_active_faults.fetch_sub(1, std::memory_order_relaxed);
  }
```

I missed this in my review. Do you or Craig have cycles to fix this? If not, I can attempt to fix it later this week (I don't really know how the new promise mechanism work, I can try to play with it).

---

Then, why this didn't break master? My theory is that if we configure `shard_count=50` to the xds e2e test, the MaxFault test cases were separated into multiple "shard"s. Each shard owns a state of gRPC library, so the global variable `g_active_faults` isn't shared between them. And unfortunately, the test case I wrote originally only validated the incremental behavior of the MaxFault counter.

I will enhance the test case to cover the decremental behavior of the MaxFault counter (after this PR is merged, or feel free to just do it).

**Comment by markdroth** [comment]
I don't have cycles to work on this right now.  If @ctiller can fix this when he gets back from vacation, that's fine.  Otherwise, I'd appreciate it if you could take a look.

Thanks!

**Comment by markdroth** [comment]
@ctiller We reverted #29288, because we realized that this would actually cause a bug in the behavior, and we needed to quickly back-port a fix to the 1.46 branch.  Hopefully it won't be too hard to roll forward once you get back.

@lidizheng Please work with Craig to fix the tests to cover this bug.

I'll merge this PR once the tests are green.

**Comment by lidizheng** [comment]
I can update the MaxFault test case after this PR is merged.

**Pull request merged**
